File: example/xx_three/material_displace.cpp
#ifdef TWO_MODULES
module;
#include <infra/Cpp20.h>
#include <xx_three/ex.h>
module two.xxthree;
#else
#include <xx_three/xx_three.h>
#include <gfx-pbr/Api.h>
#include <gfx-obj/Api.h>
#endif

using namespace two;

EX(xx_material_displace)
{
#if UI
	UNUSED(dockbar);
	SceneViewer& viewer = ui::scene_viewer(parent);
	Scene& scene = viewer.m_scene;
#else
	static Scene scene = Scene(app.m_gfx);
	static GfxViewer viewer = GfxViewer(window, scene);
#endif

	static ImporterOBJ obj_importer(app.m_gfx);

#if UI
	ui::orbit_controls(viewer);
	//controls.enableZoom = false;
	//controls.enableDamping = true;
#endif

	constexpr int height = 500; // of camera frustum

	static float r = 0.0;

	static Material* material = nullptr;
	static Node3* light = nullptr;

	if(init)
	{
		Camera& camera = viewer.m_camera;
		camera.m_fov = 40.f; camera.m_near = 1.f; camera.m_far = 10000.f;
		camera.m_eye.z = 1500.f;

		viewer.m_viewport.m_to_gamma = true;

		//var aspect = window.innerWidth / window.innerHeight;
		//camera = new THREE.OrthographicCamera(-height * aspect, height * aspect, height, -height, 1, 10000);
		//camera.position.z = 1500;
		//scene.add(camera);

		Texture& reflection = *app.m_gfx.textures().file("cube/royal.jpg.cube");

		Zone& env = scene.m_env;
		env.m_radiance.m_texture = &reflection;
		env.m_radiance.m_ambient = rgb(0xffffff) * 0.2f;
		env.m_radiance.m_colour = rgb(0xffffff);

		// lights

		Node3& ln0 = gfx::nodes(scene).add(Node3(vec3(0.f, 0.f, 2500.f)));
		gfx::lights(scene).add(Light(ln0, LightType::Point, false, rgb(0xff0000), 0.5f, 0.f));
		light = &ln0;

		Node3& ln1 = gfx::nodes(scene).add(Node3(vec3(0.f)));
		gfx::lights(scene).add(Light(ln1, LightType::Point, false, rgb(0xff6666), 1.f, 0.f));

		Node3& ln2 = gfx::nodes(scene).add(Node3(vec3(-1000.f, 0.f, 1000.f)));
		gfx::lights(scene).add(Light(ln2, LightType::Point, false, rgb(0x0000ff), 1.f, 0.f));

		Program& three = *app.m_gfx.programs().file("pbr/three");

		Material& mat = app.m_gfx.materials().create("ninja",  [&](Material& m) {
			m.m_program = &three;
			m.m_base.m_cull_mode = CullMode::None;
			m.m_pbr.m_albedo = rgb(0x888888);
			m.m_pbr.m_metallic = 1.f;
			m.m_pbr.m_roughness = 0.4f;
			m.m_lit.m_normal = app.m_gfx.textures().file("ninja/normal.jpg");
			m.m_lit.m_occlusion = app.m_gfx.textures().file("ninja/ao.jpg");
			m.m_lit.m_displace = app.m_gfx.textures().file("ninja/displacement.jpg");
			m.m_lit.m_displace = 2.436143; // from original model
			m.m_lit.m_displace_bias = -0.428408f;
			// normalScale: vec2(1, -1), // why does the normal map require negation in this case?
		});

		// float envMapIntensity = 1.0;

		material = &mat;

		Model& model = *app.m_gfx.models().file("ninjaHead_Low"); // .obj");
		//geometry.attributes.uv2 = geometry.attributes.uv;
		//geometry.center();

		Node3& n = gfx::nodes(scene).add(Node3(vec3(0.f, -175.f, 0.f) * 25.f, ZeroQuat, vec3(25.f)));
		gfx::items(scene).add(Item(n, model, 0U, &mat));
	}

#if UI
	if(Widget* dock = ui::dockitem(dockbar, "Game", { 1U }))
	{
		Widget& sheet = ui::sheet(*dock);

		auto panel = [&](const string& name) -> Widget&
		{
			Widget& s = ui::expandbox(sheet, name.c_str());
			return ui::columns(s, { 0.3f, 0.7f });
		};

		Widget& a = panel("Material");

		ui::slider_field(a, "metalness",    material->m_pbr.m_metallic.m_value,  { 0.f, 1.f, 0.01f });
		ui::slider_field(a, "roughness",    material->m_pbr.m_roughness.m_value, { 0.f, 1.f, 0.01f });
		ui::slider_field(a, "occlusion",    material->m_lit.m_occlusion.m_value, { 0.f, 1.f, 0.01f });
		ui::slider_field(a, "displacement", material->m_lit.m_displace.m_value,  { 0.f, 5.f, 0.01f });
		ui::slider_field(a, "normal scale", material->m_lit.m_normal.m_value,    { -1.f, 1.f, 0.01f });

		ui::color_field(a, "ambient", scene.m_env.m_radiance.m_ambient);

		// envMapIntensity
	}
#endif

	light->apply(vec3(2500.f * cos(r), 2500.f * sin(r), 0.f));

	r += 0.01;
}

File: example/xx_three/two/material_displace.js
// material_displace.js

var viewer = two.ui.scene_viewer(panel);
two.ui.orbit_controls(viewer);
//controls.enableZoom = false;
//controls.enableDamping = true;

var scene = viewer.scene;

var height = 500; // of camera frustum

var r = 0.0;

if(init) {
    this.importerOBJ = new two.ImporterOBJ(app.gfx);

    var camera = viewer.camera;
    camera.fov = 40.0; camera.near = 1.0; camera.far = 10000.0;
    camera.eye.z = 1500.0;

    viewer.viewport.to_gamma = true;

    var reflection = app.gfx.textures.file('cube/royal.jpg.cube');
    
    var env = scene.env;
    env.radiance.texture = reflection;
    env.radiance.ambient = 0.2;
    env.radiance.colour = two.rgb(0xffffff);

    // lights

    var zeroq = new two.quat(new two.vec3(0.0));
    
    var ln0 = scene.nodes().add(new two.Node3(new two.vec3(0.0, 0.0, 2500.0), zeroq));
    scene.lights().add(new two.Light(ln0, two.LightType.Point, false, two.rgb(0xff0000), 0.5, 0.0));
    this.light = ln0;

    var ln1 = scene.nodes().add(new two.Node3(new two.vec3(0.0), zeroq));
    scene.lights().add(new two.Light(ln1, two.LightType.Point, false, two.rgb(0xff6666), 1.0, 0.0));

    var ln2 = scene.nodes().add(new two.Node3(new two.vec3(-1000.0, 0.0, 1000.0), zeroq));
    scene.lights().add(new two.Light(ln2, two.LightType.Point, false, two.rgb(0x0000ff), 1.0, 0.0));

    var three = app.gfx.programs.file('pbr/three');

    var mat = app.gfx.materials.create('ninja'); var m = mat;
        m.program = three;
        m.pbr.albedo.value = two.rgb(0x888888);
        m.pbr.metallic.value = 1.0;
        m.pbr.roughness.value = 0.4;
        m.lit.normal.texture = app.gfx.textures.file('ninja/normal.jpg');
        m.lit.occlusion.texture = app.gfx.textures.file('ninja/ao.jpg');
        m.lit.displace.texture = app.gfx.textures.file('ninja/displacement.jpg');
        m.lit.displace.value = 2.436143; // from original model
        m.lit.displace_bias = -0.428408;
        // normalScale: vec2(1, -1), // why does the normal map require negation in this case?

    this.material = mat;

    var model = app.gfx.models.file('ninjaHead_Low'); // .obj');
    //geometry.center();

    var n = scene.nodes().add(new two.Node3(new two.vec3(0.0, -175.0 * 25.0, 0.0), zeroq, new two.vec3(25.0)));
    scene.items().add(new two.Item(n, model, 0, mat));
}

var r = time; // * 0.1;

this.light.apply(new two.vec3(2500.0 * Math.cos(r), 2500.0 * Math.sin(r), 0.0));

//this.material.pbr.roughness.value = (Math.cos(r) + 1.0) / 2.0;
this.material.pbr.metallic.value = (Math.cos(r) + 1.0) / 2.0;
this.material.lit.displace.value = (Math.sin(r) + 1.0) / 2.0 * 5.0;

r += 0.01;
